lib/core.js

```diff
@@ -688,12 +688,16 @@ declare class Promise<+R> {
     static resolve<T>(object: Promise<T> | T): Promise<T>;
     static reject<T>(error: any): Promise<T>;
     static all<T: Iterable<mixed>>(promises: T): Promise<$TupleMap<T, typeof $await>>;
+    static allSettled<T: Iterable<mixed>>(promises: T): Promise<$TupleMap<T, typeof $settlement>>;
     static race<T, Elem: Promise<T> | T>(promises: Iterable<Elem>): Promise<T>;
 }
 
 // we use this signature when typing await expressions
 declare function $await<T>(p: Promise<T> | T): T;
 
+declare function $settlement<T>(p: Promise<T> | T): SettlementResult<T>;
+declare type SettlementResult<T> = { status: 'fulfilled', value: T } | { status: 'rejected', reason: any };
```

Comment:
I have a few requests for this line:
* These types ought to be exact -- use `{|` and `|}` instead of `{` and `}` (this will be the default soon anyway).
* This will be more usable if the type parameter is covariant, so write `SettlementResult<+T>` instead of `SettlementResult<T>`.
  * This will also force you to make at least the `value` property covariant by adding a `+` in front of it. And you might as well just make the other fields covariant too, while you're at it.
* I'd rather see this use `mixed` for the rejection reason instead of `any`

Reply:
Thank you for the substantive feedback! I've made a changes on all the items.